File: rximagepicker/src/main/java/com/qingmei2/rximagepicker/ui/BaseSystemPickerView.java
package com.qingmei2.rximagepicker.ui;

import android.Manifest;
import android.app.Fragment;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.support.v4.content.ContextCompat;

import io.reactivex.Observable;
import io.reactivex.functions.Consumer;
import io.reactivex.subjects.PublishSubject;

import static android.app.Activity.RESULT_OK;

/**
 * Created by QingMei on 2018/1/19.
 */
public abstract class BaseSystemPickerView extends Fragment {

    public static final int GALLERY_REQUEST_CODE = 100;
    public static final int CAMERA_REQUEST_CODE = 101;

    private PublishSubject<Boolean> attachedSubject = PublishSubject.create();
    ;
    private PublishSubject<Uri> publishSubject;

    private PublishSubject<Integer> canceledSubject;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setRetainInstance(true);
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        attachedSubject.onNext(true);
        attachedSubject.onComplete();
    }

    public Observable<Uri> getUriObserver() {
        publishSubject = PublishSubject.create();
        canceledSubject = PublishSubject.create();

        requestPickImage();
        return publishSubject.takeUntil(canceledSubject);
    }

    private void requestPickImage() {
        if (!isAdded()) {
            attachedSubject.subscribe(new Consumer<Boolean>() {
                @Override
                public void accept(Boolean attached) throws Exception {
                    startRequest();
                }
            });
        } else {
            startRequest();
        }
    }

    public abstract void startRequest();

    private void onImagePicked(Uri uri) {
        if (publishSubject != null) {
            publishSubject.onNext(uri);
            publishSubject.onComplete();
        }
    }

    protected boolean checkPermission() {
        if (ContextCompat.checkSelfPermission(getActivity(), Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                requestPermissions(new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE}, 0);
            }
            return false;
        } else {
            return true;
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults) {
        if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            startRequest();
        }
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (resultCode == RESULT_OK) {
            switch (requestCode) {
                case GALLERY_REQUEST_CODE:
                case CAMERA_REQUEST_CODE:
                    onImagePicked(getActivityResultUri(data));
                    break;
            }
        } else {
            canceledSubject.onNext(requestCode);
        }
    }

    public abstract Uri getActivityResultUri(Intent data);
}

File: rximagepicker/src/main/java/com/qingmei2/rximagepicker/ui/BaseSystemPickerView.kt
package com.qingmei2.rximagepicker.ui

import android.Manifest
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v4.content.ContextCompat

import com.qingmei2.rximagepicker.entity.Result
import com.qingmei2.rximagepicker.function.*

import io.reactivex.Observable
import io.reactivex.functions.Consumer
import io.reactivex.subjects.PublishSubject

import android.app.Activity.RESULT_OK

abstract class BaseSystemPickerView : Fragment() {

    private val attachedSubject = PublishSubject.create<Boolean>()

    private var publishSubject: PublishSubject<Result>? = null

    private var canceledSubject: PublishSubject<Int>? = null

    val uriObserver: Observable<Result>
        get() {
            publishSubject = PublishSubject.create()
            canceledSubject = PublishSubject.create()

            requestPickImage()
            return publishSubject!!.takeUntil(canceledSubject!!)
        }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setRetainInstance(true)
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        attachedSubject.onNext(true)
        attachedSubject.onComplete()
    }

    private fun requestPickImage() {
        if (!isAdded()) {
            attachedSubject.subscribe( { startRequest() })
        } else {
            startRequest()
        }
    }

    abstract fun startRequest()

    abstract fun getActivityResultUri(data: Intent): Uri

    private fun onImagePicked(uri: Uri) {
        if (publishSubject != null) {
            publishSubject!!.onNext(parseResultNoExtraData(uri))
            publishSubject!!.onComplete()
        }
        closure()
    }

    /**
     * When the pick image behavior ending, remove this fragment from the activity.
     */
    private fun closure() {
        val fragmentManager = getFragmentManager()
        val fragmentTransaction = fragmentManager!!.beginTransaction()
        fragmentTransaction.remove(this)
        fragmentTransaction.commit()
    }

    protected fun checkPermission(): Boolean {
        if (ContextCompat.checkSelfPermission(getActivity()!!, Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                requestPermissions(arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE), 0)
            }
            return false
        } else {
            return true
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        if (grantResults.size > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            startRequest()
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent) {
        if (resultCode == RESULT_OK) {
            when (requestCode) {
                GALLERY_REQUEST_CODE, CAMERA_REQUEST_CODE -> onImagePicked(getActivityResultUri(data))
            }
        } else {
            canceledSubject!!.onNext(requestCode)
        }
    }

    companion object {

        val GALLERY_REQUEST_CODE = 100
        val CAMERA_REQUEST_CODE = 101
    }
}
